### ISSUE TITLE
Segfault & invalid free with a GRUCell

### ISSUE THREAD

**Issue by Zenol** [pytorch/pytorch#6176]
Hello,

I noticed a crash when using cuda (python segfault with an invalid free) when using GRUCell with the following invalid arguments:

* Making a GruCell.
* Feed the cell with a batch of size B > 1 for the input layer and a hidden layer of batch size 1. `self.gru_cell(input_layer, hidden_layer)`

There is a mismatch in the batch dimension of the two inputs of the model. The expected behavior should be an error message telling a mismatch in dimension. The observed behavior is a code running while using CPU implementation, and a crash with the cuda implementation.

Observed with Pytorch version 0.3.1. The bug should be easily reproducible.

Best regards,

**Comment by zou3519** [comment]
This has been fixed on master:
```
In [6]: import torch
   ...: input_size = 3
   ...: hidden_size = 4
   ...: model = nn.GRUCell(input_size, hidden_size, False).cuda()
   ...: x = torch.randn(5, input_size).cuda()
   ...: h = torch.randn(2, hidden_size).cuda()
   ...: model(x, h)
   ...:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-8b15dfb6b3f5> in <module>()
      5 x = torch.randn(5, input_size).cuda()
      6 h = torch.randn(2, hidden_size).cuda()
----> 7 model(x, h)

~/pytorch/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    369             result = self._slow_forward(*input, **kwargs)
    370         else:
--> 371             result = self.forward(*input, **kwargs)
    372         for hook in self._forward_hooks.values():
    373             hook_result = hook(self, input, result)

~/pytorch/torch/nn/modules/rnn.py in forward(self, input, hx)
    754     def forward(self, input, hx):
    755         self.check_forward_input(input)
--> 756         self.check_forward_hidden(input, hx)
    757         return self._backend.GRUCell(
    758             input, hx,

~/pytorch/torch/nn/modules/rnn.py in check_forward_hidden(self, input, hx, hidden_label
)
    507             raise RuntimeError(
    508                 "Input batch size {} doesn't match hidden{} batch size {}".form
at(
--> 509                     input.size(0), hidden_label, hx.size(0)))
    510
    511         if hx.size(1) != self.hidden_size:

RuntimeError: Input batch size 5 doesn't match hidden batch size 2
```
@Zenol, I'm not sure if this is exactly how you're triggering your error. If you could provide a code snippet to confirm that would be great.

**Comment by Zenol** [comment]
Yes, that's exactly this :)

(Your code doesn't run on my version of pytorch because he ask for Variable instead of tensors. The following lines
```
import torch
import torch.nn as nn
from torch.autograd import Variable
input_size = 3
hidden_size = 4
model = nn.GRUCell(input_size, hidden_size, False).cuda()
x = Variable(torch.randn(5, input_size).cuda())
h = Variable(torch.randn(2, hidden_size).cuda())
model(x, h)
```
produce the `free(): invalid pointer: 0x00007f78cc574d40` error.)

Looking at your output it seams fixed :)

**Comment by apaszke** [comment]
It’s fixed in master and that’s the usual free error that appears in our latest binaries. Should be fixed once 0.4 is out. Thanks for the report!
